### PR TITLE
EOS 18763 - Implementation of wait mechanism to reduce CPU load in libfabric

### PR DESCRIPTION
**Pull request by upendrapatwardhan**

### CODE REVIEW

File: net/libfab/libfab.c [hunk 1]
```diff
@@ -438,15 +442,15 @@ static void libfab_tm_event_post(struct m0_fab__tm *tm,
...
+	struct m0_net_transfer_mc *net = ftm->ftm_ntm;
 	int                        i;
 	m0_time_t                  now = m0_time_now();
 
 	M0_PRE(libfab_tm_invariant(ftm));
 	for (i = 0; i < ARRAY_SIZE(net->ntm_q); ++i) {
 		struct m0_net_buffer *nb;
 
-		m0_tl_for(m0_net_tm, &ftm->ftm_net_ma->ntm_q[i], nb) {
+		m0_tl_for(m0_net_tm, &ftm->ftm_ntm->ntm_q[i], nb) {
```
Comment (huanghua78) [inline net/libfab/libfab.c:453]: s/&ftm->ftm_ntm->ntm_q[i]/&net->ntm_q[i]/

Comment (huanghua78) [reply net/libfab/libfab.c:453]: To iterate a list, some lock must be held.
We must take a lock here, or assert this lock is already held.

File: net/libfab/libfab.c [hunk 2]
```diff
@@ -545,22 +549,55 @@ static void libfab_txep_event_check(struct m0_fab__ep *txep,
...
 	if (cq != NULL) {
-		rc = libfab_check_for_comp(cq, &buf, &len);
+		rc = libfab_check_for_comp(cq, buf, len);
 		if (rc > 0) {
-			if (buf->fb_length == 0)
-				buf->fb_length = len;
-			buf->fb_ev_ep = ep;
-			libfab_ep_get(ep);
-			libfab_buf_done(buf, 0);
+			for (i = 0; i < rc; i ++) {
```
Comment (huanghua78) [inline net/libfab/libfab.c:566]: No space before the "self increment operator".
```
for (i = 0; i < rc; i++) {
```

File: net/libfab/libfab.c (hunk 1 above)
Comment (upendrapatwardhan) [reply net/libfab/libfab.c]: Done.

**Review by jamohite — APPROVED** [review]

File: net/libfab/libfab.c [hunk 3]
```diff
@@ -545,22 +551,55 @@ static void libfab_txep_event_check(struct m0_fab__ep *txep,
...
 	if (cq != NULL) {
-		rc = libfab_check_for_comp(cq, &buf, &len);
+		rc = libfab_check_for_comp(cq, buf, len);
 		if (rc > 0) {
-			if (buf->fb_length == 0)
-				buf->fb_length = len;
-			buf->fb_ev_ep = ep;
-			libfab_ep_get(ep);
-			libfab_buf_done(buf, 0);
+			for (i = 0; i < rc; i++) {
```
Comment (huanghua78) [inline net/libfab/libfab.c:568]: The condition "if (rc > 0)" is covered by "for (i = 0; i < rc; i++)".
Can we drop the "if (rc > 0"?

File: net/libfab/libfab.c [hunk 4]
```diff
@@ -545,22 +551,55 @@ static void libfab_txep_event_check(struct m0_fab__ep *txep,
...
+	struct m0_fab__ep        *xep;
+	struct m0_fab__active_ep *aep;
+	struct m0_fab__buf       *buf[FAB_MAX_COMP_READ];
+	int                       i;
+	int                       rc;
+
+	memset(&buf, 0, sizeof(buf));
+	rc = libfab_check_for_comp(cq, buf, NULL);
+	if (rc > 0) {
+		for (i = 0; i < rc; i++) {
```
Comment (huanghua78) [inline net/libfab/libfab.c:595]: The same as the previous comment.

File: net/libfab/libfab.c [hunk 5]
```diff
@@ -1309,34 +1349,24 @@ static int libfab_tm_param_free(struct m0_fab__tm *tm)
...
-	if (tm->ftm_waitset != NULL)
-		return M0_RC(rc);
+	M0_PRE(tm->ftm_epfd == -1);
 
-	memset(&wait_attr, 0, sizeof(wait_attr));
-	wait_attr.wait_obj = FI_WAIT_UNSPEC;
-	rc = fi_wait_open(tm->ftm_fab->fab_fab, &wait_attr, &tm->ftm_waitset);
+	tm->ftm_epfd = epoll_create(1);
+	if (tm->ftm_epfd < 0)
+		return -errno;
```
Comment (huanghua78) [inline net/libfab/libfab.c:1367]: return M0_ERR(-errno);

File: net/libfab/libfab.c [hunk 6]
```diff
@@ -1873,6 +1906,22 @@ static int libfab_txep_init(struct m0_fab__active_ep *aep,
...
 
+static int libfab_waitfd_bind(struct fid* fid, struct m0_fab__tm *tm)
+{
+	struct epoll_event ev;
+	int                fd;
+	int                rc;
+
+	rc = fi_control(fid, FI_GETWAIT, &fd);
+	if (rc != FI_SUCCESS)
+		return M0_RC(rc);
```
Comment (huanghua78) [inline net/libfab/libfab.c:1917]: return M0_ERR(rc);

File: net/libfab/libfab_internal.h [hunk 7]
```diff
@@ -67,34 +66,61 @@ enum m0_fab__mr_params {
...
+/**
+ * Libfab structure of fabric params
+ */
 struct m0_fab__fab {
-	uint64_t           fab_magic;          /* Magic number */
-	struct fi_info    *fab_fi;             /* Fabric interface info */
-	struct fid_fabric *fab_fab;            /* Fabric fid */
-	struct fid_domain *fab_dom;            /* Domain fid */
-	struct m0_tlink    fab_link;           /* List of fabrics */
+
```
Comment (huanghua78) [inline net/libfab/libfab_internal.h:95]: Remove this blank line.

File: net/libfab/libfab_internal.h [hunk 8]
```diff
@@ -67,34 +66,61 @@ enum m0_fab__mr_params {
...
+	struct m0_tlink    fab_link;
 };
 
+/**
+ * Libfab structure of endpoint name
+ */
 struct m0_fab__ep_name {
-	char fen_addr[LIBFAB_ADDR_LEN_MAX];    /*  */
-	char fen_port[LIBFAB_PORT_LEN_MAX];    /* Port range 0-65535 */
+
```
Comment (huanghua78) [inline net/libfab/libfab_internal.h:116]: Remove this blank line.

Comment (huanghua78) [reply net/libfab/libfab_internal.h:116]: and similar issues for other struct.

File: net/libfab/libfab.c (hunk 3 above)
Comment (upendrapatwardhan) [reply net/libfab/libfab.c]: Done

File: net/libfab/libfab.c (hunk 4 above)
Comment (upendrapatwardhan) [reply net/libfab/libfab.c]: Done

File: net/libfab/libfab.c (hunk 5 above)
Comment (upendrapatwardhan) [reply net/libfab/libfab.c]: Done

File: net/libfab/libfab.c (hunk 6 above)
Comment (upendrapatwardhan) [reply net/libfab/libfab.c]: Done

File: net/libfab/libfab_internal.h (hunk 7 above)
Comment (upendrapatwardhan) [reply net/libfab/libfab_internal.h]: Done

File: net/libfab/libfab_internal.h (hunk 8 above)
Comment (upendrapatwardhan) [reply net/libfab/libfab_internal.h]: Done

**Review by jamohite — APPROVED** [review]

File: net/libfab/libfab.c [hunk 9]
```diff
@@ -542,26 +546,55 @@ static void libfab_txep_event_check(struct m0_fab__ep *txep,
...
-			if (rc > 0) {
-				if (buf->fb_length == 0)
-					buf->fb_length = len;
-				buf->fb_ev_ep = ep;
-				libfab_ep_get(ep);
-				libfab_buf_done(buf, 0);
+		rc = libfab_check_for_comp(cq, buf, len);
+		if (rc > 0) {
+			for (i = 0; i < rc; i ++) {
+				if (buf[i] != NULL) {
```
Comment (sg-shankar) [inline net/libfab/libfab.c]: Check if we can add assert if buf[i] is null

File: net/libfab/libfab.c [hunk 10]
```diff
@@ -542,26 +546,55 @@ static void libfab_txep_event_check(struct m0_fab__ep *txep,
...
+	struct m0_fab__active_ep *aep;
+	struct m0_fab__buf       *buf[FAB_MAX_COMP_READ];
+	int                       i;
+	int                       rc;
+
+	memset(&buf, 0, sizeof(buf));
+	rc = libfab_check_for_comp(cq, buf, NULL);
+	if (rc > 0) {
+		for (i = 0; i < rc; i ++) {
+			if (buf[i] != NULL) {
```
Comment (sg-shankar) [inline net/libfab/libfab.c]: Check if we can add assert if buf[i] is null

**Review by sg-shankar — APPROVED** [review]
Discussed the comment with Upendra and Jaydeep

**Pull request merged**